src/orion/service/client/workon.py

```diff
@@ -0,0 +1,178 @@
+        result = self._post(
+            "suggest", experiment_name=experiment_name, pool_size=pool_size
+        )
+
+        trials = []
+        for trial in result["trials"]:
+            trials.append(
+                RemoteTrial(**trial, exp_working_dir=self.experiment.working_dir)
+            )
+
+        # Currently we only return a single trial
+        trial = trials[0]
+
+        # Create pacemaker for them
+        self._maintain_reservation(trial)
+        return TrialCM(self, trial)
+
+    def _maintain_reservation(self, trial):
+        self._pacemakers[trial.id] = TrialPacemaker(trial, self)
+        self._pacemakers[trial.id].start()
+
+    def release(self, trial, status="interrupted"):
+        """Release a trial after work on it finished"""
+        experiment_name = self.experiment_name
+
+        if experiment_name is None:
+            raise ExperiementIsNotSetup("experiment_name is not set")
+
+        failed = False
+        try:
+            self._post(
+                "release",
+                euid=self.experiment_id,
+                experiment_name=experiment_name,
+                trial_hash=trial.id,
+                status=status,
+            )
+        except Exception as e:
+            failed = True
+            raise e
+        finally:
+            pacemaker = self._pacemakers.pop(trial.id, None)
+            if pacemaker is not None:
+                pacemaker.stop()
+            elif not failed:
+                raise RuntimeError(
+                    f"Trial {trial.id} had no pacemakers. Was it reserved properly?"
+                )
+
+    def observe(
+        self,
+        trial: RemoteTrial,
+        results: List[Dict],
+        name: str = "objective",
+        experiment_name=None,
+    ) -> None:
+        """Observe the result of a given trial"""
+        experiment_name = experiment_name or self.experiment_name
+
+        if isinstance(results, numbers.Number):
+            results = [dict(value=results, name=name, type="objective")]
+
+        if experiment_name is None:
+            raise ExperiementIsNotSetup("experiment_name is not set")
+
+        if not isinstance(results, list):
+            raise TypeError("Results need to be a float or a list of values")
+
+        self._post(
+            "observe",
+            euid=self.experiment_id,
+            experiment_name=experiment_name,
+            trial_id=trial.db_id,
+            trial_hash=trial.params_id,
+            results=results,
+        )
+
+        pacemaker = self._pacemakers.pop(trial.id, None)
+        if pacemaker is not None:
+            pacemaker.stop()
+        else:
+            log.warning("Trial pacemaker not found! %s", trial.id)
+
+    def is_done(self) -> bool:
+        """returns true if the experiment is done"""
+        payload = self._post(
+            "is_done", experiment_name=self.experiment_name, euid=self.experiment_id
+        )
+        return payload.get("is_done", True)
```

Comment:
What could cause the `payload` to not contain `is_done`? Would these situations mean that the experiment is indeed done?

Reply:
that is just defensive programing